File: src/ast.cc
#include <ast.h>
#include <stdexcept>

Expr* parse_expr(Parser* p);

static IfStmt parse_if(Parser* p);

static IfStmt parse_if(Parser* p) {
    std::vector<CondBlock> branches;
    std::optional<Block> else_body;

    auto cond = parse_expr(p);
    if (!p->at(TK_LBRACE)) {
        throw std::runtime_error("expected '{' after if condition");
    }
    p->advance();

    std::vector<Stmt> stmts;
    while (!p->at(TK_RBRACE)) {
        extern Stmt parse_stmt(Parser*);
        stmts.push_back(parse_stmt(p));
    }

    p->advance();
    branches.push_back(CondBlock{ cond, Block{ stmts } });

    while (p->at(TK_ELIF)) {
        p->advance();
        auto elif_cond = parse_expr(p);
        if (!p->at(TK_LBRACE)) {
            throw std::runtime_error("expected '{' after elif condition");
        }

        p->advance();
        std::vector<Stmt> elif_stmts;
        while (!p->at(TK_RBRACE)) {
            extern Stmt parse_stmt(Parser*);
            elif_stmts.push_back(parse_stmt(p));
        }

        p->advance();
        branches.push_back(CondBlock{ elif_cond, Block{ elif_stmts } });
    }

    if (p->at(TK_ELSE)) {
        p->advance();
        if (!p->at(TK_LBRACE)) {
            throw std::runtime_error("expected '{' after else");
        }

        p->advance();
        std::vector<Stmt> else_stmts;
        while (!p->at(TK_RBRACE)) {
            extern Stmt parse_stmt(Parser*);
            else_stmts.push_back(parse_stmt(p));
        }

        p->advance();
        else_body = Block{ else_stmts };
    }

    return IfStmt{ branches, else_body };
}

Stmt parse_stmt(Parser* p) {
    if (p->at(TK_LET)) {
        p->advance();
        auto ident = std::get<std::string>(p->advance().val.data);
        if (p->at(TK_ASSIGN)) {
            p->advance();
        } else {
            throw std::runtime_error("expected '=' after let binding name");
        }

        auto* expr = parse_expr(p);

        p->expect(TK_EOS);
        return LetStmt{ ident, expr };

    } else if (p->at(TK_IDENT) && p->peek(1).tk == TK_ASSIGN) {
        auto ident = std::get<std::string>(p->advance().val.data);
        p->advance();

        auto* expr = parse_expr(p);
        p->expect(TK_EOS);

        return AssignStmt{ ident, expr };

    } else if (p->at(TK_RET)) {
        p->advance();
        auto* expr = parse_expr(p);
        p->expect(TK_EOS);

        return RetStmt{ expr };

    } else if (p->at(TK_IF)) {
        p->advance();
        return parse_if(p);

    } else if (p->at(TK_FN)) {
        p->advance();
        auto ident = std::get<std::string>(p->advance().val.data);

        if (!p->at(TK_LPAREN)) {
            throw std::runtime_error("expected '(' after function name");
        }
        p->advance();

        std::vector<std::string> args;
        while (!p->at(TK_RPAREN)) {
            if (p->at(TK_IDENT)) {
                args.push_back(std::get<std::string>(p->advance().val.data));
            } else {
                throw std::runtime_error("expected identifier in parameter list");
            }
            if (p->at(TK_COMMA)) {
                p->advance();
            }
        }
        p->advance();

        if (!p->at(TK_LBRACE)) {
            throw std::runtime_error("expected '{' for function body");
        }
        p->advance();

        std::vector<Stmt> stmts;
        while (!p->at(TK_RBRACE)) {
            stmts.push_back(parse_stmt(p));
        }
        p->advance();

        return FnDeclStmt{ ident, args, Block{ stmts } };

    } else {
        auto* expr = parse_expr(p);
        p->expect(TK_EOS);
        return ExprStmt{ expr };
    }
}

std::vector<Stmt> parse(std::vector<Token> tks) {
    auto psr = Parser{ tks };
    std::vector<Stmt> stmts;
    while (!psr.at_end()) {
        stmts.push_back(parse_stmt(&psr));
    }
    return stmts;
}

File: src/ast.rs
pub use crate::error::ParserError;
pub use crate::tokenizer::Token;
pub use crate::value::Value;

#[derive(Clone)]
pub enum Expr {
    Binary {
        op: Token,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Unary {
        op: Token,
        arg: Box<Expr>,
    },
    Call {
        ident: String,
        args: Vec<Expr>,
    },
    Index {
        expr: Box<Expr>,
        idx: Box<Expr>,
    },
    Ident {
        ident: String,
    },
    Literal {
        val: Value,
    },
    ArrayLiteral {
        members: Vec<Expr>,
    },
}

pub type Block = Vec<Stmt>;

#[derive(Clone)]
pub struct CondBlock {
    pub cond: Expr,
    pub body: Block,
}

#[derive(Clone)]
pub struct  LetStmt {
    pub name: String,
    pub expr: Expr
}

#[derive(Clone)]
pub struct  AssignStmt {
    pub name: String,
    pub expr: Expr
}

#[derive(Clone)]
pub struct  RetStmt {
    pub expr: Expr
}

#[derive(Clone)]
pub struct  IfStmt {
    pub branches: Vec<CondBlock>,
    pub else_body: Option<Block>,
}

#[derive(Clone)]
pub struct  WhileStmt {
    pub block: CondBlock
}

#[derive(Clone)]
pub struct  FnDeclStmt {
    pub name: String,
    pub args: Vec<String>,
    pub body: Block,
}

#[derive(Clone)]
pub struct  ExprStmt {
    pub expr: Expr,
}

#[derive(Clone)]
pub struct  ForInStmt {
    pub lhs: Expr,
    pub rhs: Expr,
    pub block: Block
}

#[derive(Clone)]
pub struct  ForICMStmt {
    pub init: Box<Stmt>,
    pub cond: Expr,
    pub fmod: Box<Stmt>,
    pub block: Block
}

#[derive(Clone)]
pub enum Stmt {
    Assign(AssignStmt),
    Let(LetStmt),
    Ret(RetStmt),
    If(IfStmt),
    While(WhileStmt),
    FnDecl(FnDeclStmt),
    Expr(ExprStmt),
    ForIn(ForInStmt),
    ForICM(ForICMStmt), // init; cond; mod
    Continue,
    Break
}

pub struct Parser {
    pub tokens: Vec<Token>,
    pub pos: usize,
    use_eos: bool
}

impl Parser {
    pub fn at_end(&self) -> bool {
        self.pos >= self.tokens.len() || self.tokens[self.pos] == Token::Eof
    }

    pub fn peek(&self) -> Token {
        if self.at_end() {
            return Token::Eof;
        }
        self.tokens[self.pos].clone()
    }

    pub fn peek_n(&self, n: usize) -> Token {
        if self.pos + n >= self.tokens.len() {
            return Token::Eof;
        }
        self.tokens[self.pos + n].clone()
    }

    pub fn advance(&mut self) -> Token {
        if self.at_end() {
            return Token::Eof;
        }
        let tk = self.tokens[self.pos].clone();
        self.pos += 1;
        tk
    }

    pub fn advance_n(&mut self, n: usize) -> Token {
        if self.pos + n >= self.tokens.len() {
            return Token::Eof;
        }
        let tk = self.tokens[self.pos].clone();
        self.pos += n;
        tk
    }

    pub fn at(&self, t: Token) -> bool {
        self.peek() == t
    }

    pub fn expect(&mut self, t: Token) -> Result<Token, ParserError> {
        if !self.at(t.clone()) {
            Err(ParserError::UnexpectedToken{expected: t, got: self.peek()})
        } else {
            Ok(self.advance())
        }
    }

    pub fn enable_eos(&mut self, yes: bool) {
        self.use_eos = yes;
    }

    pub fn ensure_eos(&mut self) -> Result<Token, ParserError> {
        if self.use_eos {
            self.expect(Token::Eos)
        } else {
            Ok(Token::Eos)
        }
    }

    fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, pos: 0, use_eos: true }
    }
}

pub fn parse(toks: Vec<Token>) -> Result<Vec<Stmt>, ParserError> {
    let mut psr = Parser::new(toks);
    let mut statements = Vec::<Stmt>::new();
    while !psr.at_end() {
        statements.push(crate::ast_stmt::parse_stmt(&mut psr)?);
    }
    Ok(statements)
}
